src/worker/mod.rs

```diff
@@ -197,18 +197,27 @@ impl Worker {
     /// completed
     pub(crate) async fn shutdown(&self) {
         let _ = self.shutdown_sender.send(true);
+        // First, we want to stop polling of both activity and workflow tasks
         if let Some(atm) = self.at_task_mgr.as_ref() {
             atm.notify_shutdown();
         }
-        self.wf_task_poll_buffer.notify_shutdown();
+        self.wf_task_source.stop_pollers();
+        // Next we need to wait for all local activities to finish so no more workflow task
+        // heartbeats will be generated
+        self.local_act_mgr.wait_all_finished().await;
+        // Then we need to wait for any tasks generated as a result of completing WFTs, which
+        // heartbeating generates
+        self.wf_task_source
+            .wait_for_tasks_from_complete_to_drain()
+            .await;
         // Notify in case shutdown was requested while there were no more outstanding WFTs.
         // This is required because the only other place where we notify wfts_drained is on
         // activation completion and activation polling checks for wfts_drained.
         self.maybe_notify_wtfs_drained();
         // wait until all outstanding workflow tasks have been completed before shutting down
-        if !*self.wfts_drained.borrow() {
-            self.wfts_drained
-                .clone()
+        let mut drained = self.wfts_drained.clone();
+        while !*drained.borrow() {
```

Comment:
Impossible to prevent without doing some nasty stuff like inspecting if a WFT completion is starting local activities and then somehow writing down that we shouldn't consider that particular WFT as complete for this purpose (but it is actually complete) etc etc. Local activities are just ugly that way.